cpp/src/endpoint.cpp

```diff
@@ -566,15 +576,36 @@ std::shared_ptr<Worker> Endpoint::getWorker() { return ::ucxx::getWorker(_parent
 void Endpoint::errorCallback(void* arg, ucp_ep_h ep, ucs_status_t status)
 {
   ErrorCallbackData* data = reinterpret_cast<ErrorCallbackData*>(arg);
+
+  // Unable to cast to `ErrorCallbackData*`: invalid `arg`.
+  if (data == nullptr) {
+    ucxx_error("ucxx::Endpoint::%s, UCP handle: %p, error callback called with status %d: %s",
+               __func__,
+               ep,
+               status,
+               ucs_status_string(status));
+    return;
+  }
+
+  std::shared_ptr<Endpoint> endpoint{nullptr};
+  try {
+    endpoint = data->endpoint.lock();
+  } catch (std::bad_weak_ptr& exception) {
+    // Unable to acquire `std::shared_ptr<ucxx::Endpoint>`: owner was already destroyed.
+    return;
```

Comment:
*issue*: [`std::weak_ptr::lock`](https://en.cppreference.com/w/cpp/memory/weak_ptr/lock) is `noexcept`, so this error handling isn't doing what you expect.

You probably want:
```c++
if (!(endpoint = data.endpoint.lock())) {
  return;
}
```

i.e. attempt to upgrade to a `shared_ptr` and return if that is unsuccesful.

Reply:
Thanks for catching that. This solution works too, but I think the "proper" C++ solution is `try...catch` with the `weak_ptr` being passed to the `shared_ptr` constructor, so I've done that in https://github.com/rapidsai/ucxx/pull/227/commits/454762c0a03a5b2642f24fdeaa6115d6ec85696a.